src/connector/src/sink/mongodb.rs

```diff
@@ -665,66 +635,61 @@ impl MongodbPayloadWriter {
         match op {
             Op::Insert | Op::UpdateInsert => upsert_builder
                 .entry(ns)
-                .or_insert_with(|| UpsertCommandBuilder::new(coll, self.max_entries))
+                .or_insert_with(|| UpsertCommandBuilder::new(coll))
                 .add_upsert(pk, document)?,
             Op::UpdateDelete => (),
             Op::Delete => upsert_builder
                 .entry(ns)
-                .or_insert_with(|| UpsertCommandBuilder::new(coll, self.max_entries))
+                .or_insert_with(|| UpsertCommandBuilder::new(coll))
                 .add_delete(pk)?,
         }
-
-        self.buffered_entries += 1;
-        if self.buffered_entries >= self.max_entries {
-            self.flush_upsert(upsert_builder).await?;
-        }
         Ok(())
     }
 
-    async fn flush_insert(
-        &mut self,
+    fn flush_insert(
+        &self,
         insert_builder: &mut HashMap<MongodbNamespace, InsertCommandBuilder>,
-    ) -> Result<()> {
+    ) -> Result<Vec<impl futures::Future<Output = std::result::Result<(), SinkError>>>> {
         // TODO try sending bulk-write of each collection concurrently to improve the performance when
         // `dynamic collection` is enabled. We may need to provide best practice to guide user on setting
         // the MongoDB driver's connection properties.
+        let mut futures = Vec::with_capacity(insert_builder.len());
         for (ns, builder) in insert_builder.drain() {
-            self.send_bulk_write_command(&ns.0, builder.build()).await?;
+            let db = self.client.database(&ns.0);
+            futures.push(Self::send_bulk_write_command(db, builder.build()));
         }
-        self.buffered_entries = 0;
-        Ok(())
+        Ok(futures)
     }
 
-    async fn flush_upsert(
-        &mut self,
+    fn flush_upsert(
+        &self,
         upsert_builder: &mut HashMap<MongodbNamespace, UpsertCommandBuilder>,
-    ) -> Result<()> {
+    ) -> Result<Vec<impl futures::Future<Output = std::result::Result<(), SinkError>>>> {
```

Comment:
We can change to return try_join_all(...).boxed() here so that we can avoid allocating an unnecessary intermediate vec.

We can collect the future by converting the builder into an iterator of future by code like the following
```
try_join_all(upsert_builder.into_iter().flat_map(|(ns, builder)| {
    let (upsert, delete) = builder.build();
    let db = self.client.database(&ns.0);
    upsert
        .map(|upsert| Self::send_bulk_write_command(db.clone(), upsert))
        .into_iter()
        .chain(delete.map(|delete| Self::send_bulk_write_command(db, delete)))
}))
```

So as `flush_insert`.